### ISSUE TITLE
[AMDGPU] `mul` used with v1i8 / v1i16 causes crash during IR optimizations due to type mismatch

### ISSUE THREAD

**Issue by HazyFish** [llvm/llvm-project#58331]
## Description

For `amdgcn`, when `mul` is used with `v1i8` / `v1i16` causes crash during IR optimizations due to type mismatch.
The problem doesn't exist for `v1i32`.

## Cause

In function [`AMDGPUCodeGenPrepare::replaceMulWithMul24`](https://github.com/llvm/llvm-project/blob/12c6a41f5249fa97041754d38cac28877ed49611/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp#L509), 

https://github.com/llvm/llvm-project/blob/12c6a41f5249fa97041754d38cac28877ed49611/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp#L570-L573

`insertValues` will turn a vector of size 1 into a scalar due to the following logic:

https://github.com/llvm/llvm-project/blob/12c6a41f5249fa97041754d38cac28877ed49611/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp#L468-L472

As a result, when `replaceAllUsesWith` is called, the assertion `New->getType() == getType() && "replaceAllUses of value with new value of different type!"` failed because the old type (vector) and the new type (scaler) are not the same.

## Minimal Reproduction

https://godbolt.org/z/48n93sTWf

### Code

```ll
define <1 x i16> @f(<1 x i16> %0) {
BB:
  %B = mul <1 x i16> %0, <i16 42>
  ret <1 x i16> %B
}
```

### Stack Trace

```console
llc: /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/Value.cpp:506: void llvm::Value::doRAUW(llvm::Value *, llvm::Value::ReplaceMetadataUses): Assertion `New->getType() == getType() && "replaceAllUses of value with new value of different type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: ./llvm-project/build-debug/bin/llc ./crash-reports/dagisel-amdgcn/1.ll -mtriple=amdgcn
1.	Running pass 'Function Pass Manager' on module './crash-reports/dagisel-amdgcn/1.ll'.
2.	Running pass 'AMDGPU IR optimizations' on function '@f'
 #0 0x0000000003af78fa llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:569:11
 #1 0x0000000003af7aab PrintStackTraceSignalHandler(void*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:636:1
 #2 0x0000000003af60f6 llvm::sys::RunSignalHandlers() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Signals.cpp:104:5
 #3 0x0000000003af81d5 SignalHandler(int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007fe6d0720980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #5 0x00007fe6cf610e87 raise /build/glibc-uZu3wS/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #6 0x00007fe6cf6127f1 abort /build/glibc-uZu3wS/glibc-2.27/stdlib/abort.c:81:0
 #7 0x00007fe6cf6023fa __assert_fail_base /build/glibc-uZu3wS/glibc-2.27/assert/assert.c:89:0
 #8 0x00007fe6cf602472 (/lib/x86_64-linux-gnu/libc.so.6+0x30472)
 #9 0x00000000030596bd llvm::Value::doRAUW(llvm::Value*, llvm::Value::ReplaceMetadataUses) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/Value.cpp:0:3
#10 0x0000000003059cd2 llvm::Value::replaceAllUsesWith(llvm::Value*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/Value.cpp:534:1
#11 0x0000000001b19503 (anonymous namespace)::AMDGPUCodeGenPrepare::replaceMulWithMul24(llvm::BinaryOperator&) const /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:573:3
#12 0x0000000001b17cdd (anonymous namespace)::AMDGPUCodeGenPrepare::visitBinaryOperator(llvm::BinaryOperator&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:1240:7
#13 0x0000000001b167bd llvm::InstVisitor<(anonymous namespace)::AMDGPUCodeGenPrepare, bool>::visitMul(llvm::BinaryOperator&) /home/henry/aflplusplus-isel/llvm-project/llvm/include/llvm/IR/Instruction.def:151:1
#14 0x0000000001b1600c llvm::InstVisitor<(anonymous namespace)::AMDGPUCodeGenPrepare, bool>::visit(llvm::Instruction&) /home/henry/aflplusplus-isel/llvm-project/llvm/include/llvm/IR/Instruction.def:151:1
#15 0x0000000001b15cf2 (anonymous namespace)::AMDGPUCodeGenPrepare::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:1440:21
#16 0x0000000002f995d6 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1430:23
#17 0x0000000002f9e402 llvm::FPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1476:16
#18 0x0000000002f99ea9 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1545:23
#19 0x0000000002f99a1d llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:535:16
#20 0x0000000002f9e6e1 llvm::legacy::PassManager::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1672:3
#21 0x0000000000d391ec compileModule(char**, llvm::LLVMContext&) /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:737:41
#22 0x0000000000d37592 main /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:418:13
#23 0x00007fe6cf5f3c87 __libc_start_main /build/glibc-uZu3wS/glibc-2.27/csu/../csu/libc-start.c:344:0
#24 0x0000000000d36d9a _start (./llvm-project/build-debug/bin/llc+0xd36d9a)
```

**Comment by arsenm** [comment]
https://reviews.llvm.org/D135821

**Comment by arsenm** [comment]
Fixed by 838fd611b79e3514eead5dd724140df046172ef1
